lib/client.ml

```diff
@@ -95,6 +130,171 @@ let route client router msg =
+    sequence_number
+
+  let wait_ack t sequence_number =
+    let cond =
+      Base.Hashtbl.find_or_add t.acknowledges sequence_number
+        ~default:Lwt_condition.create in
+    Lwt_condition.wait cond
+
+  let wait_ack_timeout t sequence_number timeout =
+    Lwt.pick
+      [
+        (let _ = Lwt_unix.sleep (Float.of_int timeout) in
+         Lwt.return @@ Result.Error "Timeout reached");
+        (let _ = wait_ack t sequence_number in
+         Lwt.return @@ Result.Ok "Successfully received acknowledge");
+      ]
+
+  (** Basic random shuffle, see https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle*)
+  let knuth_shuffle known_peers =
+    let shuffled_array = Array.copy (Array.of_list known_peers) in
+    let initial_array_length = Array.length shuffled_array in
+    for i = initial_array_length - 1 downto 1 do
+      let k = Random.int (i + 1) in
+      let x = shuffled_array.(k) in
+      shuffled_array.(k) <- shuffled_array.(i);
+      shuffled_array.(i) <- x
+    done;
+    Array.to_list shuffled_array
+
+  (* Regarding the SWIM protocol, if peer A cannot get ACK from peer B (timeout):
+     A sets B as `suspicious`
+     A randomly picks one (or several, should it also be randomly determined?) peer(s) from its list
+     and ask him/them to ping B.*)
+  (* This function return the random peer, to which we will ask to ping the first peer *)
+  let rec pick_random_neighbors neighbors number_of_neighbors =
+    let addresses = neighbors |> Base.Hashtbl.keys |> knuth_shuffle in
+    match addresses with
+    | [] -> failwith "pick_random_peers"
+    | elem :: _ ->
+      if number_of_neighbors = 1 then
+        [elem]
+      else
+        elem :: pick_random_neighbors neighbors (number_of_neighbors - 1)
+
+  let update_neighbor_status peer neighbor status =
+    let open Peer in
+    let local_neighbor = get_neighbor peer neighbor.address in
+    match local_neighbor with
+    | Some neighbor ->
+      neighbor.status <- status;
+      Base.Hashtbl.update peer.neighbors neighbor.address ~f:(fun _ -> neighbor)
+    | None -> ()
+  let create_message client message (recipient : Peer.t) =
+    Message.
+      {
+        category = Failure_detection;
+        id = -1;
+        sender = address_of !client;
+        recipient = recipient.address;
+        payload = Util.Encoding.pack bin_writer_message message;
+      }
+
+  let send_message message client (recipient : Peer.t) =
+    let message = create_message client message recipient in
+    send_to client message
+
+  let send_ping_to client peer = send_message Ping client peer
+
+  let send_acknowledge_to client peer = send_message Acknowledge client peer
+
+  let send_ping_request_to client (recipient : Peer.t) =
+    send_message (PingRequest recipient.address) client recipient
+
+  (** Processes an incoming message bound for the failure detector of a client *)
+  let handle_message client message =
+    let open Message in
+    let peer_of_client = peer_from !client in
```

Comment:
Since there is no `where syntax` in OCaml (shame!), is it possible to get the value right before needing it?
I think it is easier to avoid question like
> ok, where does `peer_of_client` comes from already?

What do you think?